hibernate-core/src/main/java/org/hibernate/persister/collection/AbstractCollectionPersister.java

```diff
@@ -363,6 +362,8 @@ public AbstractCollectionPersister(
 		else {
 			elementPersister = null;
 		}
+		// Defer this after the element persister was determined, because it is needed in OneToManyPersister#getTableName()
+		spaces[0] = getTableName();
```

Comment:
This changes the entry for the space 0 from e.g. `AbstractTablePerClassEntity` to `(select ... from table1 union all select ... from table2 ...)` for one-to-many associations, to match what the entity persister reports as its table name. Needed for some assert statements.

Relatedly, wouldn't it be more correct to copy the spaces of the element entity type into the collection spaces? AFAIU, the spaces affect cache invalidation, and in case of an abstract table per class entity element, we wouldn't really invalidate any caches though we certainly should.